File: wc2a.c
#include <ctype.h>
#include <stdlib.h>
#include <stdio.h>

#include <unistd.h>
#include <fcntl.h>

enum {BUFSIZE=65536};

struct results {
    unsigned long long byte_count;
    unsigned long long word_count;
    unsigned long long line_count;
};

/* ASCII spaces = " \t\f\v\n\r" */
static unsigned char is_space[] = {
    0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
};

/** Parse a single 64k chunk. Since a word can cross a chunk
 * boundary, we have to remember the 'state' from a previous
 * chunk. */
void parse_chunk(const char *buf, size_t length, struct results *results, int *in_state)
{
    /* Setup optimized variables */
    enum { Spaces=0, Word=1} state = *in_state;
    size_t i = 0;
    unsigned long long line_count = results->line_count;
    unsigned long long word_count = results->word_count;

    /* Run the inner loop. This is where 99.9% of the time is spent
     * in this program. */
    while (i < length) {
        char c = buf[i++];
        int is_not_space = !is_space[(unsigned char)c];

        line_count += (c == '\n');
        word_count += ((state == Spaces) & is_not_space);

        state = is_not_space;
    }

    /* Save the optimized variables */
    *in_state = state;
    results->line_count = line_count;
    results->word_count = word_count;
}

/** Parse an individual file, or <stdin>, and print the results */
void parse_file(char *buf, size_t bufsize, int fd, const char *filename)
{
    struct results results = {0, 0, 0};
    int state = 0; /* state held between chunks */

    for (;;) {
        ssize_t count;

        /* Process a 64k chunk at a time */
        count = read(fd, buf, bufsize);
        if (count <= 0)
            break;
        results.byte_count += count;
        parse_chunk(buf, count, &results, &state);
    }

    printf("%8llu %7llu %7llu %s\n", results.line_count, results.word_count, results.byte_count, filename);
}

int main(int argc, char *argv[])
{
    char *buf;
    
    /* Allocate a buffer. We don't read in the entire file, but
     * one small bit at a time. */
    buf = malloc(BUFSIZE);
    if (buf == NULL)
        abort();


    if (argc == 1) {
        /* If no files specified, then process <stdin> instead */
        parse_file(buf, BUFSIZE, STDIN_FILENO, "");
    } else {
        int i;

        /* Process all the files specified on the command-line */
        for (i=1; i<argc; i++) {
            int fd;

            fd = open(argv[1], O_RDONLY);
            if (fd == -1) {
                perror(argv[1]);
                continue;
            }

            parse_file(buf, BUFSIZE, fd, argv[1]);

            close(fd);
        }
    }

    return 0;
}

File: wc2a.js
var fs = require("fs");

var table = new Array();

function Results()
{
    this.line_count = 0;
    this.word_count = 0;
    this.char_count = 0;
    this.was_space = true;
    this.update = function (line_count, word_count, char_count, was_space) {
        this.line_count += line_count;
        this.word_count += word_count;
        this.char_count += char_count;
        this.was_space = was_space;
        return this;
    }
    this.update2 = function (s) {
        return this.update(s.line_count, s.word_count, s.char_count, s.was_space);
    }
}
function isspace(c) {
    return (9 <= c && c <= 13) || (c == 32);
}

function parse_chunk(buf, length, state)
{
    var was_space = state.was_space;
    var line_count = 0;
    var word_count = 0;
    var c;
    var is_space;
    var i;

    for (i=0; i<length; i++) {
        c = buf[i];
        is_space = table[c];
        line_count += (c == 10) ? 1 : 0;
        word_count += (!is_space && was_space) ? 1 : 0;
        was_space = is_space;
    }

    return state.update(line_count, word_count, length, was_space);
}

function print_results(results, cfg, filename)
{
    var str = "";
    if (cfg.is_line_count)
        str = str + " " + results.line_count;
    if (cfg.is_word_count)
        str = str + " " + results.word_count;
    if (cfg.is_char_count)
        str = str + " " + results.char_count;
    if (filename && filename.length > 0)
        str = str + " " + filename;
    console.log(str);
}

function parse_file(fd, filename, cfg)
{
    var state = new Results();
    var buf = Buffer.alloc(65536);

    for (;;) {
        var length = fs.readSync(fd, buf, 0, 65536);
        if (length == 0)
            break;
        state = parse_chunk(buf, length, state);
    }
    
    print_results(state, cfg, filename);
    return state;
}


function print_help()
{
    console.log("usage: wc [-clmw] [file ...]");
    process.exit(1);
}

function parse_configuration(argv)
{
    var cfg = {is_line_count:false, is_word_count:false, is_char_count:false};
    
    for (var i in argv) {
        if (argv[i].charAt(0) == '-') {
            for (var j=1; j<argv[i].length; j++) {
                switch (argv[i].charAt(j)) {
                    case 'l': cfg.is_line_count = true; break;
                    case 'w': cfg.is_word_count = true; break;
                    case 'c': cfg.is_char_count = true; break;
                    case '?': print_help(); break;
                    case 'h': print_help(); break;
                    case '-':
                        if (argv[i].charAt(j+1) == 'h')
                            print_help();
                        else {
                            console.log("[-] unknown option: " + argv[i]);
                            process.exit(1);
                        }
                        break;
                    default:
                        console.log("[-] unknown option: -" + argv[i].charAt(i));
                        process.exit(1);
                        break;
                }
            }
        }
    }

    if (cfg.is_line_count == false && cfg.is_word_count == false && cfg.is_char_count == false)
        cfg = {is_line_count: true, is_word_count: true, is_char_count: true};

    return cfg;
}

function main(argv)
{
    var file_count = 0;
    var total = new Results();
    
    /* setup character table to determine spaces */
    for (var i=0; i<256; i++)
        table[i] = isspace(i);

    /* Parse configuration */
    var cfg = parse_configuration(argv);

    /* Parse all the files */
    for (var i in argv) {
        var filename = argv[i];
        if (i < 2)
            continue;
        if (argv[i].length == 0)
            continue;
        if (argv[i].charAt(0) == '-')
            continue;
        try {
            var fd = fs.openSync(filename, "r");
            total.update2(parse_file(fd, filename, cfg));
            fs.closeSync(fd);
            file_count++;
        } catch (err) {
            if (err.syscall == 'open' && err.code == 'ENOENT') {
                console.log("[-] " + err.path + ": No such file or directory");
            } else 
                console.log(err);
            process.exit(1);
            break;
        }
    }
    if (file_count == 0)
        total.update2(parse_file(0, "", cfg));
    if (file_count > 1)
        print_results(total, cfg, "total");

}

main(process.argv);

